transformer_engine/pytorch/module/base.py

```diff
@@ -1238,15 +1252,19 @@ def get_weight_workspace(
         workspace_dtype: torch.dtype, default = None
             If weight workspace contains high-precision tensor - for example
             for debug quantization, this is dtype of the tensor.
+
         """
 
-        # FP8 primary weights
+        # Handle case where weights are already quantized
+        # Note: Make sure weights have required usages, but do not
+        # destroy unnecessary usages since they may be used later.
         if isinstance(tensor, QuantizedTensor):
-            if update_workspace and quantizer is not None:
-                tensor.update_usage(
-                    rowwise_usage=quantizer.rowwise_usage,
-                    columnwise_usage=quantizer.columnwise_usage,
-                )
+            update_rowwise_usage = True if quantizer.rowwise_usage else None
+            update_columnwise_usage = True if quantizer.columnwise_usage else None
+            tensor.update_usage(
+                rowwise_usage=update_rowwise_usage,
+                columnwise_usage=update_columnwise_usage,
+            )
```

Comment:
Destroying unnecessary usages causes problems when alternating between training steps (column-wise data needed) and validation steps (column-wise data not needed).  See https://github.com/NVIDIA/TransformerEngine/pull/1832#discussion_r2113016499.

Reply:
TBH this issue is just because of optimizer not doing the right job with quantizing. If we made it so it uses the quantizer then we would not need this part at all.

Reply:
The problem is fundamental to validation steps. The validation steps don't require column-wise data, so we were aggressively deallocating to save memory. Once we started more training steps, we failed in bprop because we were missing column-wise data.